### DISCUSSION TITLE
Running two independent ProviderContainers "side-by-side"

### DISCUSSION THREAD

**Discussion by derolf** [rrousselGit/riverpod#2494] [General]
I have a use-cases, where I basically show two apps side-by-side (client/server apps connected via IP). 

Both have their own "world" and hence I created a parentless ProviderContainer for both. However, there are some issues around vsync here since only one container can own the vsyncOverride.

So, I manually defined my own vsyncOverride:

```Dart
  // ignore: invalid_use_of_internal_member
  vsyncOverride = (void Function() task) {
    WidgetsBinding.instance.scheduleFrameCallback((_) => task());
  };
```

I am wondering why you are not using `WidgetsBinding.instance.scheduleFrameCallback` in your `_flutterVsync`, but doing this:

```
   if (SchedulerBinding.instance.schedulerPhase ==
        SchedulerPhase.transientCallbacks) {
      markNeedsBuild();
    } else {
      // Using microtask, otherwise Flutter tests complain about pending timers
      Future.microtask(() {
        if (_mounted) markNeedsBuild();
      });
    }
 ```

**Comment by rrousselGit** [comment]
>  However, there are some issues around vsync here since only one container can own the vsyncOverride.

What issue?

**Reply by derolf** [reply]
I saw these:

- Sometimes I get “Only one task can be scheduled at a time”
- Sometimes only one of the two apps updates on provider changes

**Reply by rrousselGit** [reply]
I see. You should file bug reports for these then.

**Reply by derolf** [reply]
I am still wondering why you are not using ‘WidgetsBinding.instance.scheduleFrameCallback((_) => task());’.

**Reply by dishank-betaflux** [reply]
I am facing the same issue? Still not quite sure what is the root cause of it. It is not affecting me with anything major. getting exception on the debugger means something is going agains the rules.

error msg : Only one task can be scheduled at a time.

**Reply by tgrushka** [reply]
Yes, I am getting this error message “Only one task can be scheduled at a time” constantly, all the time, using StreamProvider.

I wish I could be more specific, and I could, if only the error were more specific. It's very frustrating because I'm just converting a stream into a provider so I can use it with Riverpod, but the documentation on StreamProvider, and how to use ref.listen, ref.watch vs ref.read within the StreamProvider, or async* vs sync, or whatever, is the right way. They all compile and run just fine until trying to consume them, then constantly this error  “Only one task can be scheduled at a time”, which is totally and utterly meaningless.

And I am only using one provider container in my app, so I doubt this bug is even related to using more than one container.

**Reply by penguisnt** [reply]
I am seeing the same issue using StreamProvider, but I'm not using side-by-side, just a regular app with a yield* to a Firestore `.snapshots` call.

It seems to arise after some hot-refresh stuff, and a provider gets abandonded somewhere (even with autodispose), because I can see the logs from the old StreamProvider still present

**Reply by rrousselGit** [reply]
Please give a way to reproduce the issue

**Comment by josh-burton** [comment]
I have the exact same situation where I'm running multiple apps within a parent app (its a storybook type app that renders my app multiple times in different variations).

The workaround of overriding vsync works, but would be great to have a proper fix since this API is scheduled to be removed in 3.0.

**Reply by rrousselGit** [reply]
Please give a way to reproduce the issue

**Comment by runionke** [comment]
I've also been having the same message with StreamProvider on firebase RTDB snapshots. It seems to occur after a hot restart has occurred.

**Reply by rrousselGit** [reply]
Please give a way to reproduce the issue

**Comment by josh-burton** [comment]
@rrousselGit  The issue still exists for me, and now that vsyncOverride has been removed the workaround no longer works.

I've created a test that reproduces the issue: https://github.com/rrousselGit/riverpod/issues/3118

**Reply by abel-123** [reply]
@rrousselGit any update on this? 
I have the same issue/ error message, although for me the issue appears when calling Update/ Set to some documents that are also being streamed.
